### PR TITLE
Scattered cleanups

### LINKED ISSUE #369: Can't use `OverloadedRecordDot` when extracting the `"data"` field from `PTxInfo`
### Problem
`data` is a reserved word in Haskell and hence you can't do something like `txInfo.data` without the Parser complaining. 

### Solution
Perhaps rename to `data'` or `dataMap`?

### LINKED ISSUE #400: Fix remaining issues in #326

### LINKED ISSUE #402: Unexport `hrecField`
We already have `getField` which does the same thing.

### PR DESCRIPTION
**Pull request by TotallyNotChase**
* Unexport `hrecField`. Resolves #402 
* Revert removal of `PValue` and `PMap` constructors. Resolves #400 
* Rename `PTxInfo`'s `data` field to `datums`. Resolves #369 
* Add `Num` instance for `PPOSIXTime` and export its constructor.
* Change `PIsDataRepr` superclass to be `PlutusType`, not `PMatch`.

Also trying to fix a few more small things that will be pushed up later.

### CODE REVIEW

**Comment by L-as** [comment]
Merge in staging to fix CI

**Comment by TotallyNotChase** [comment]
* Change `HRec`'s kind to `[(Symbol, Type)] -> Type` - Resolves #385 

I've been thinking of refining this one further. It's possible to do-
```hs
type Labeled :: (Symbol, Type) -> Type
newtype Labeled labeledX = Labeled {unLabeled :: Snd labeledX}

type HRec :: [(Symbol, Type)] -> Type
type HRec as = NP Labeled as
```
which is beautiful! But I don't know if a type alias is the right thing to do here since we have instances for `HRec`.

More importantly, I _believe_ the most elegant type for `HRec` would involve `[(Symbol, PType)]` as the type level list parameter - but I think that'll make quite a few classes and families a fair bit more complicated. We could even have the type level list be `[(Symbol, k)]` and make `HRec` (and everything else) take a `f :: k -> Type` (generics-sop style). I don't know if any of these refinements are worth the trouble though.

**Review by L-as — APPROVED** [review]
Looks good to me for now.

**Comment by L-as** [comment]
> * Change `HRec`'s kind to `[(Symbol, Type)] -> Type` - Resolves [Fix kind of HRec #385](https://github.com/Plutonomicon/plutarch/issues/385)
> 
> 
> I've been thinking of refining this one further. It's possible to do-
> 
> ```haskell
> type Labeled :: (Symbol, Type) -> Type
> newtype Labeled labeledX = Labeled {unLabeled :: Snd labeledX}
> 
> type HRec :: [(Symbol, Type)] -> Type
> type HRec as = NP Labeled as
> ```
> 
> which is beautiful! But I don't know if a type alias is the right thing to do here since we have instances for `HRec`.
> 
> More importantly, I _believe_ the most elegant type for `HRec` would involve `[(Symbol, PType)]` as the type level list parameter - but I think that'll make quite a few classes and families a fair bit more complicated. We could even have the type level list be `[(Symbol, k)]` and make `HRec` (and everything else) take a `f :: k -> Type` (generics-sop style). I don't know if any of these refinements are worth the trouble though.

You could just make it a newtype, but that would ruin the point a bit and would complicate constructing and destructing.
I think it's fine the way you've done it.

**Comment by TotallyNotChase** [comment]
The last thing I tried to work on is #392. There's a WIP of it in the latest commit but I can't see how the `pmatch` could be implemented in a useful way - does anyone have any ideas?

File: Plutarch/DataRepr/Internal.hs [hunk 1]
```diff
@@ -108,6 +107,12 @@ data PDataRecord (as :: [PLabeledType]) (s :: S) where
     PDataRecord ((name ':= x) ': xs) s
   PDNil :: PDataRecord '[] s
 
+instance {-# OVERLAPPABLE #-} PlutusType (PDataRecord l) where
+  type PInner (PDataRecord l) _ = PBuiltinList PData
+  pcon' (PDCons x xs) = pcons # pforgetData x # pto xs
+  pcon' PDNil         = pnil
+  pmatch' _ _ = undefined -- FIXME
```
Comment (L-as) [inline Plutarch/DataRepr/Internal.hs:114]: You could just do `error "you need a more specific PDataRecord"` lol

Comment (TotallyNotChase) [reply Plutarch/DataRepr/Internal.hs:114]: that's exactly what I've been leaning towards after trying out a bunch of things that didn't work....

that might honestly be the best thing to do right now. This instance is only really useful for the `pto` anyway.

File: Plutarch/DataRepr/Internal.hs [hunk 2]
```diff
@@ -108,6 +107,12 @@ data PDataRecord (as :: [PLabeledType]) (s :: S) where
     PDataRecord ((name ':= x) ': xs) s
   PDNil :: PDataRecord '[] s
 
+instance {-# OVERLAPPABLE #-} PlutusType (PDataRecord l) where
+  type PInner (PDataRecord l) _ = PBuiltinList PData
+  pcon' (PDCons x xs) = pcons # pforgetData x # pto xs
+  pcon' PDNil         = pnil
```
Comment (L-as) [inline Plutarch/DataRepr/Internal.hs:113]: these should probably also err

Comment (L-as) [reply Plutarch/DataRepr/Internal.hs:113]: or maybe not, doesn't seem like they are any less efficient? anyway, I think you can actually do `pcon' (PDCons x xs) = pcon' (PDCons x xs)` and have it work

**Review by L-as — CHANGES_REQUESTED** [review]

File: Plutarch/DataRepr/Internal.hs (hunk 2 above)
Comment (L-as) [reply Plutarch/DataRepr/Internal.hs:113]: You should probably add type annotations to make it clear what's happening

File: Plutarch/DataRepr/Internal.hs (hunk 1 above)
Comment (L-as) [reply Plutarch/DataRepr/Internal.hs:114]: in a dependently typed language we could pattern match on the `l` probably

Comment (TotallyNotChase) [reply Plutarch/DataRepr/Internal.hs:114]: it's ok I heard GHC is landing `DependentHaskell` tomorrow

File: Plutarch/DataRepr/Internal.hs (hunk 2 above)
Comment (TotallyNotChase) [reply Plutarch/DataRepr/Internal.hs]: added type annotations, also made it use the underlying `pcon'` instead - good idea!

**Comment by TotallyNotChase** [comment]
'tis ready

**Review by L-as — APPROVED** [review]
LGTM!

**Pull request merged**